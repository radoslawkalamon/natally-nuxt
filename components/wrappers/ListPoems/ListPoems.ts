import Vue from 'vue'
import { DTOMetaPostPoemConstructor, DTOMetaPostPoem } from '@/utils/dto.meta.post.poem'

export default Vue.extend({
  name: 'WrappersListPoems',
  props: {
    limit: {
      type: Number,
      default: 0
    },
    randomize: {
      type: Boolean,
      default: false
    }
  },
  data (): Record<string, DTOMetaPostPoemConstructor[]> {
    return {
      poems: []
    }
  },
  async fetch (): Promise<void> {
    this.poems = this.randomize
      ? await this.fetchMetaPostPoemsRandom()
      : await this.fetchMetaPostPoems()
  },
  computed: {
    fetchContentDirectory (): string {
      return '/poezja-314'
    },
    metaPostPoemsConstructorKeys (): string[] {
      return ['createdAt', 'description', 'imageCover', 'imageOpenGraph', 'path', 'title', 'updatedAt']
    },
    metaPostPoems (): DTOMetaPostPoem[] {
      return this.poems.map((i: DTOMetaPostPoemConstructor) => new DTOMetaPostPoem(i))
    }
  },
  methods: {
    async fetchMetaPostPoems (): Promise<DTOMetaPostPoemConstructor[]> {
      const poems = await this.$content(this.fetchContentDirectory)
        .only(this.metaPostPoemsConstructorKeys)
        .sortBy('createdAt', 'desc')
        .limit(this.limit)
        .fetch<DTOMetaPostPoemConstructor>()

      return !Array.isArray(poems)
        ? [poems]
        : poems
    },
    async fetchMetaPostPoemsRandom (): Promise<DTOMetaPostPoemConstructor[]> {
      const poems = await this.$content(this.fetchContentDirectory)
        .only(this.metaPostPoemsConstructorKeys)
        .fetch<DTOMetaPostPoemConstructor>()

      return !Array.isArray(poems)
        ? [poems]
        : poems
          .sort(() => 0.5 - Math.random())
          .slice(0, this.limit || undefined)
    }
  }
})
