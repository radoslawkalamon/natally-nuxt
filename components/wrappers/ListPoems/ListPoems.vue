<template>
  <div class="list-poems">
    <slot v-bind="{ metaPostPoems }" />
  </div>
</template>

<script src="./ListPoems.ts" lang="ts"></script>
