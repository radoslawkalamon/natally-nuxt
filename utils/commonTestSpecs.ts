import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { merge } from 'lodash'

export function shallRender (component: typeof Vue, dataTestID: string, defaultOptions?: object): void {
  test('shall render', (): void => {
    const wrapper = mount(component, defaultOptions)
    const isWrapperVisible = wrapper.find(`[data-test="${dataTestID}"]`).isVisible()
    expect(isWrapperVisible).toBeTruthy()
  })
}

export function shallHaveSlot (component: typeof Vue, slot: string, defaultOptions?: object): void {
  test(`shall have slot '${slot}'`, (): void => {
    const text = `This is slot '${slot}' test.`
    const wrapperOptions = merge(defaultOptions, {
      slots: {
        [slot]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHaveStringProp (component: typeof Vue, propName: string, defaultOptions?: object): void {
  test(`shall have string prop '${propName}'`, (): void => {
    const text = `This is prop ${propName} test.`
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHavePropInfluenceOverClassNames (
  component: typeof Vue,
  propName: string,
  propValue: unknown,
  classNameSuffix: string,
  defaultOptions?: object
) {
  test(`shall have ${typeof propValue} prop '${propName}' influence over class names`, () => {
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: propValue
      }
    })
    const wrapper = mount(component, wrapperOptions)
    const isWrapperHaveClass = wrapper.classes().some(cls => cls.includes(classNameSuffix))
    expect(isWrapperHaveClass).toBeTruthy()
  })
}
