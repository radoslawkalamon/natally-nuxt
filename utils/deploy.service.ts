/* eslint-disable no-console */
import { Client, FileInfo, FTPResponse } from 'basic-ftp'
import { ProgressInfo } from 'basic-ftp/dist/ProgressTracker'
import capitalize from './formatters/capitalize'
import spacelize from './formatters/spacelize'

type DeployServiceConfig = {
  host: string;
  port: number;
  username: string;
  password: string;
  remoteCatalog: string;
  remoteIgnoreFile: string[];
  localCatalog: string;
}

type RemoveStrategies = {
  [key in FileInfo['type']]: (name: FileInfo['name']) => Promise<void | FTPResponse>
}

function decoratorLogCall () {
  return function (_target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const targetMethod = descriptor.value

    descriptor.value = function (...args: any[]) {
      const formattedName = spacelize(capitalize(propertyKey))
      console.log(`::: ${formattedName} :::`)
      return targetMethod.apply(this, args)
    }

    return descriptor
  }
}

class DeployService {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly remoteCatalog: string;
  readonly remoteIgnoreFile: string[];
  readonly localCatalog: string;
  FTPClient = new Client()

  constructor (config: DeployServiceConfig) {
    this.host = config.host
    this.port = Number(config.port)
    this.username = config.username
    this.password = config.password
    this.remoteCatalog = config.remoteCatalog
    this.remoteIgnoreFile = config.remoteIgnoreFile
    this.localCatalog = config.localCatalog
  }

  public async start (): Promise<void> {
    try {
      await this.openConnection()
      await this.ensureRemoteCatalog()
      await this.openRemoteCatalog()
      await this.cleanRemoteCatalog()
      await this.uploadToRemoteCatalog()
    } catch (e) {
      console.error(e)
      throw e
    } finally {
      this.closeConnection()
    }
  }

  @decoratorLogCall()
  private async openConnection (): Promise<void> {
    await this.FTPClient.access({
      host: this.host,
      port: this.port,
      user: this.username,
      password: this.password,
      secure: false
    })
  }

  @decoratorLogCall()
  private async ensureRemoteCatalog (): Promise<void> {
    await this.FTPClient.ensureDir(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async openRemoteCatalog (): Promise<void> {
    await this.FTPClient.cd(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async cleanRemoteCatalog (): Promise<void> {
    const remoteFiles = await this.FTPClient.list()

    for (const file of remoteFiles) {
      if (this.remoteIgnoreFile.includes(file.name)) {
        continue
      }

      const removeStrategies: RemoveStrategies = {
        0: async () => await Promise.resolve(),
        1: async name => await this.FTPClient.remove(name),
        2: async name => await this.FTPClient.removeDir(name),
        3: async () => await Promise.resolve()
      }
      await removeStrategies[file.type](file.name)
    }
  }

  @decoratorLogCall()
  private async uploadToRemoteCatalog (): Promise<void> {
    this.FTPClient.trackProgress(this.handlerUploadProgress)
    await this.FTPClient.uploadFromDir(this.localCatalog)
    this.FTPClient.trackProgress()
  }

  private handlerUploadProgress (info: ProgressInfo): void {
    if (info.bytes === 0) {
      process.stdout.write(`Uploading file ${info.name}\n`)
    }
  }

  @decoratorLogCall()
  private closeConnection (): void {
    this.FTPClient.close()
  }
}

export default DeployService
