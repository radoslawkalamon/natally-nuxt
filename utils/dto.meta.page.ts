export type DTOMetaPageConstructor = {
  createdAt: string;
  description: string;
  imageOpenGraph?: string;
  path: string;
  title: string;
  updatedAt: string;
}

const getPath = (path: string) => {
  const pathWithoutIndex = path.split('/').filter(i => i !== 'index').join('/')
  return `${pathWithoutIndex}${pathWithoutIndex.at(-1) !== '/' ? '/' : ''}`
}

export class DTOMetaPage {
  readonly createdAt: Date;
  readonly description: string;
  readonly imageOpenGraph: string;
  readonly path: string;
  readonly title: string;
  readonly updatedAt: Date;

  constructor ({
    createdAt,
    description,
    imageOpenGraph = 'common.opengraph.png',
    title,
    updatedAt,
    path
  }: DTOMetaPageConstructor) {
    this.createdAt = new Date(createdAt)
    this.description = description
    this.imageOpenGraph = `/images/${imageOpenGraph}`
    this.path = getPath(path)
    this.title = title
    this.updatedAt = new Date(updatedAt)
  }
}
