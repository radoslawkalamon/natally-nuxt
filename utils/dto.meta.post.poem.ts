import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostPoemConstructor = DTOMetaPageConstructor & {
  imageCover: string
  imageOpenGraph: string
}

export class DTOMetaPostPoem extends DTOMetaPage {
  readonly imageCover: string;
  readonly imageOpenGraph: string;

  constructor ({
    createdAt,
    description,
    imageCover,
    imageOpenGraph,
    path,
    title,
    updatedAt
  }: DTOMetaPostPoemConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
  }

  get imagePath (): string {
    return '/images/poezja-314/'
  }
}
