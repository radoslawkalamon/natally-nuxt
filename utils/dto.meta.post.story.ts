import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostStoryConstructor = DTOMetaPageConstructor & {
  audiobookId?: string;
  imageCover: string;
  imageCover2x: string;
  imageOpenGraph: string
  timeReading: number;
}

export class DTOMetaPostStory extends DTOMetaPage {
  readonly audiobookId: string;
  readonly imageCover: string;
  readonly imageCover2x: string;
  readonly imageOpenGraph: string;
  readonly timeReading: number;

  constructor ({
    audiobookId = '0',
    createdAt,
    description,
    imageCover,
    imageCover2x,
    imageOpenGraph,
    path,
    timeReading,
    title,
    updatedAt
  }: DTOMetaPostStoryConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.audiobookId = audiobookId
    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageCover2x = `${this.imagePath}${imageCover2x}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
    this.timeReading = timeReading
  }

  get imagePath (): string {
    return '/images/'
  }
}
