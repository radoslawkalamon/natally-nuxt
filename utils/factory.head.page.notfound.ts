import type { MetaInfo } from 'vue-meta'
import { factoryHeadSchemaOrganization } from './factory.head.schema.organization'
import { factoryHeadSchemaWebsite } from './factory.head.schema.website'
import { getTitle } from './factory.head.utils'

const title = 'Strony nie znaleziono'
const description = 'Porobiło się! Niestety strona, której szukasz nie istnieje!'

export const factoryHeadPageNotfound = (): MetaInfo => ({
  title: getTitle({ title }),
  meta: [
    { name: 'robots', content: 'noindex, follow', hid: 'meta:robots' },
    { name: 'description', content: description, hid: 'meta:description' },
    { property: 'og:title', content: getTitle({ title }), hid: 'meta:og:title' },
    { property: 'og:description', content: description, hid: 'meta:og:description' }
  ],
  script: [{
    type: 'application/ld+json',
    json: {
      '@context': 'http://schema.org',
      '@graph': [
        factoryHeadSchemaOrganization(),
        factoryHeadSchemaWebsite()
      ]
    }
  }]
})
