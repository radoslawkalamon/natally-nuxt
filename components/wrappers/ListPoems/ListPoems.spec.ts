import ListPoems from '@/components/wrappers/ListPoems/ListPoems.vue'
import { shallPassUnitSanityTest } from '@/devtools/jest.common.spec'
import { createDefaultOptionsFactory, createUnitTestWrapper } from '@/devtools/jest.common.spec.utils'
import { JestMockNuxtContent } from '@/devtools/jest.mock.nuxt.content'

jest.mock('@/utils/dto.meta.post.poem', () => ({
  DTOMetaPostPoem: class DTOMetaPostPoem {}
}))

const defaultOptionsFactory = createDefaultOptionsFactory({
  mocks: {
    $content: JestMockNuxtContent(Array(10).fill({}))
  },
  slots: {
    default: 'This is slot "default" test.'
  }
})

describe('Wrappers / List Poems', () => {
  describe('Unit', () => {
    shallPassUnitSanityTest({
      component: ListPoems,
      options: defaultOptionsFactory()
    })

    describe('Randomize: false', () => {
      test('shall { metaPostPoems } match snapshot', async () => {
        const wrapper = await createUnitTestWrapper({
          component: ListPoems,
          options: defaultOptionsFactory()
        })
        // @ts-ignore
        expect(wrapper.vm.metaPostPoems).toMatchSnapshot()
        wrapper.destroy()
      })

      test('shall $content.limit() be called with 3 when limit = 3', async () => {
        const MockNuxtContent = JestMockNuxtContent({})
        const wrapper = await createUnitTestWrapper({
          component: ListPoems,
          options: defaultOptionsFactory({
            mocks: {
              $content: MockNuxtContent
            },
            propsData: {
              limit: 3
            }
          })
        })
        expect(MockNuxtContent().limit).toHaveBeenCalledWith(3)
        wrapper.destroy()
      })

      test('shall $content.where() be called with "/mock-path-false/" when without = "/mock-path-false/"', async () => {
        const MockNuxtContent = JestMockNuxtContent({})
        const wrapper = await createUnitTestWrapper({
          component: ListPoems,
          options: defaultOptionsFactory({
            mocks: {
              $content: MockNuxtContent
            },
            propsData: {
              without: ['/mock-path-false/']
            }
          })
        })
        expect(MockNuxtContent().where).toHaveBeenCalledWith({ path: { $nin: ['/mock-path-false/'] } })
        wrapper.destroy()
      })
    })

    describe('Randomize: true', () => {
      test('shall { metaPostPoems } match snapshot', async () => {
        const wrapper = await createUnitTestWrapper({
          component: ListPoems,
          options: defaultOptionsFactory({
            propsData: {
              randomize: true
            }
          })
        })
        // @ts-ignore
        expect(wrapper.vm.metaPostPoems).toMatchSnapshot()
        wrapper.destroy()
      })

      test('shall { metaPostPoems } have length 3 when limit = 3', async () => {
        const wrapper = await createUnitTestWrapper({
          component: ListPoems,
          options: defaultOptionsFactory({
            propsData: {
              limit: 3,
              randomize: true
            }
          })
        })
        // @ts-ignore
        expect(wrapper.vm.metaPostPoems).toHaveLength(3)
        wrapper.destroy()
      })

      test('shall $content.where() be called with "/mock-path-true/" when without = "/mock-path-true/"', async () => {
        const MockNuxtContent = JestMockNuxtContent({})
        const wrapper = await createUnitTestWrapper({
          component: ListPoems,
          options: defaultOptionsFactory({
            mocks: {
              $content: MockNuxtContent
            },
            propsData: {
              randomize: true,
              without: ['/mock-path-true/']
            }
          })
        })
        expect(MockNuxtContent().where).toHaveBeenCalledWith({ path: { $nin: ['/mock-path-true/'] } })
        wrapper.destroy()
      })
    })
  })
})
