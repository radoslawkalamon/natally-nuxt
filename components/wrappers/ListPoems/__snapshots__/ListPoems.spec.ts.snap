// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrappers / List Poems Unit Randomize: false shall { metaPostPoems } match snapshot 1`] = `
Array [
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
]
`;

exports[`Wrappers / List Poems Unit Randomize: true shall { metaPostPoems } match snapshot 1`] = `
Array [
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
  DTOMetaPostPoem {},
]
`;

exports[`Wrappers / List Poems Unit shall pass unit sanity test 1`] = `"<div class=\\"list-poems\\">This is slot \\"default\\" test.</div>"`;
